File: packages/ocp-crypto/src/utils/hash.ts
import { sha512 } from "@noble/hashes/sha512";
import { u32le, utf8 } from "./bytes.js";
import { scalarFromBytesModOrder } from "./scalar.js";
import type { Scalar } from "./scalar.js";

const HASH_TO_SCALAR_PREFIX = utf8("OCPv1|hash_to_scalar|");

export function hashToScalar(domainSep: string, ...msgs: Uint8Array[]): Scalar {
  const h = sha512.create();
  h.update(HASH_TO_SCALAR_PREFIX);

  const dst = utf8(domainSep);
  h.update(u32le(dst.length));
  h.update(dst);

  for (const m of msgs) {
    if (!(m instanceof Uint8Array)) throw new Error("hashToScalar: expected bytes");
    h.update(u32le(m.length));
    h.update(m);
  }

  const digest = h.digest(); // 64 bytes
  // Interpret as little-endian integer and reduce mod q.
  return scalarFromBytesModOrder(digest);
}

File: apps/chain/internal/ocpcrypto/hash.go
package ocpcrypto

import (
	"crypto/sha512"
	"fmt"
	"hash"
)

var (
	hashToScalarPrefix = []byte("OCPv1|hash_to_scalar|")
)

func updateLenBytes(h hash.Hash, b []byte) {
	h.Write(u32le(uint32(len(b))))
	h.Write(b)
}

// HashToScalar implements docs/INTERFACES.md 4.1 and WS3 hashToScalar.
func HashToScalar(domainSep string, msgs ...[]byte) (Scalar, error) {
	h := sha512.New()
	h.Write(hashToScalarPrefix)
	updateLenBytes(h, []byte(domainSep))
	for _, m := range msgs {
		if m == nil {
			return Scalar{}, fmt.Errorf("hashToScalar: nil msg")
		}
		updateLenBytes(h, m)
	}
	digest := h.Sum(nil) // 64 bytes
	return ScalarFromUniformBytes(digest)
}

